core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -514,9 +505,18 @@ private[log] class Cleaner(val id: Int,
         val retainDeletes = currentSegment.lastModified > deleteHorizonMs
         info(s"Cleaning segment $startOffset in log ${log.name} (largest timestamp ${new Date(currentSegment.largestTimestamp)}) " +
           s"into ${cleaned.baseOffset}, ${if(retainDeletes) "retaining" else "discarding"} deletes.")
-        cleanInto(log.topicPartition, currentSegment.log, cleaned, map, retainDeletes, log.config.maxMessageSize,
-          transactionMetadata, log.activeProducersWithLastSequence, stats)
 
+        try {
+          cleanInto(log.topicPartition, currentSegment.log, cleaned, map, retainDeletes, log.config.maxMessageSize,
+            transactionMetadata, log.activeProducersWithLastSequence, stats)
+        } catch {
+          case e: LogSegmentOffsetOverflowException =>
+            // KAFKA-6264: if we got a LogSegmentOverflowException, split the current segment. It's also safest to abort
```

Comment:
nit: we mention KAFKA-6264 in a few places in this patch. At least a few of these mentions can be removed if we document the problem clearly inside `LogSegmentOffsetOverflowException`.